redbeams/src/main/java/com/sequenceiq/redbeams/controller/v4/databaseserver/DatabaseServerV4Controller.java

```diff
@@ -217,4 +223,11 @@ public void start(@TenantAwareParam @ResourceCrn @ValidCrn(resource = CrnResourc
     public void stop(@TenantAwareParam @ResourceCrn @ValidCrn(resource = CrnResourceDescriptor.DATABASE_SERVER) @NotNull String crn) {
         redbeamsStopService.stopDatabaseServer(crn);
     }
+
+    @Override
+    @InternalOnly
+    public void upgrade(@TenantAwareParam @NotEmpty @ValidCrn(resource = CrnResourceDescriptor.DATABASE_SERVER) String databaseServerCrn,
+            UpgradeDatabaseServerV4Request request) {
```

Comment:
Perhaps annotate `request` with `@Valid @NotNull`.

Reply:
Is already done in the Endpoint class. Should it be done also in the Controller class?

Reply:
Maybe not. :smile: I uses to annotate in both places, but it should also be fine this way.